File: frontend/src/types/apiKeys.ts
// PARTPILOT:REST_API_KEY_SETTINGS_TYPES:V618
export type ApiKeyScope =
  | "inventory:read"
  | "inventory:write"
  | "catalogues:read"
  | "catalogues:write"
  | "projects:read"
  | "projects:write"
  | "reservations:read"
  | "reservations:write"
  | "history:read";

export type ApiKeyStatus = "active" | "expired" | "revoked";

export interface ApiKeySummary {
  id: number;
  name: string;
  masked_key: string;
  scopes: ApiKeyScope[];
  status: ApiKeyStatus;
  created_at: string;
  updated_at: string;
  expires_at: string | null;
  rotated_at: string | null;
  last_used_at: string | null;
  revoked_at: string | null;
}

export interface ApiKeyListResponse {
  keys: ApiKeySummary[];
  total: number;
  available_scopes: ApiKeyScope[];
}

export interface ApiKeySecretResponse extends ApiKeySummary {
  key: string;
}

export interface ApiKeyMutationPayload {
  name: string;
  scopes: ApiKeyScope[];
  expires_at: string | null;
}

File: backend/app/schemas/api_keys.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


# PARTPILOT:REST_API_KEY_SCHEMA:V615
ApiKeyScope = Literal[
    "inventory:read",
    "inventory:write",
    "catalogues:read",
    "catalogues:write",
    "projects:read",
    "projects:write",
    "reservations:read",
    "reservations:write",
    "history:read",
]
ApiKeyStatus = Literal["active", "expired", "revoked"]


class ApiKeyCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str = Field(min_length=1, max_length=120)
    scopes: list[ApiKeyScope] = Field(min_length=1, max_length=9)
    expires_at: datetime | None = None


class ApiKeyUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str = Field(min_length=1, max_length=120)
    scopes: list[ApiKeyScope] = Field(min_length=1, max_length=9)
    expires_at: datetime | None = None


class ApiKeySummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: int = Field(ge=1)
    name: str
    masked_key: str
    scopes: list[ApiKeyScope]
    status: ApiKeyStatus
    created_at: datetime
    updated_at: datetime
    expires_at: datetime | None
    rotated_at: datetime | None
    last_used_at: datetime | None
    revoked_at: datetime | None


class ApiKeyListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    keys: list[ApiKeySummaryResponse]
    total: int = Field(ge=0)
    available_scopes: list[ApiKeyScope]


class ApiKeySecretResponse(ApiKeySummaryResponse):
    key: str = Field(min_length=32, max_length=160, repr=False)
